src/tools/replay.cpp

```diff
@@ -62,7 +62,7 @@ void record_action(std::string action_type, std::string inner_text) {
 	log_document.SaveFile(log_file);
 }
 
-Element* pop_next_action(std::string expected_action_type) {
+std::unique_ptr<Element> pop_next_action(std::string expected_action_type) {
```

Comment:
I don't really understand why this became a `unique_ptr`. I know I said not to use raw pointers, but the pointers to XML objects are pointing to persistent memory inside the XML document, right? So I think they could just be passed around as references.

…though I guess that would assume you're not modifying the document as you go.

But is making it a `unique_ptr` like this actually safe? I don't know what `Clone()` did to allocate this element.

Reply:
It was, indeed, wrong to use `unique_ptr` and causing a crash that I couldn't figure out last night. I'm not very well versed in the standard library pointer types.

The behavior I wanted was this:

* there is one global variable, log_document, which loads all the xml at launch and unloads it at program exit
* pop_next_action() removes the next element from log_document, passing ownership of the element's data to the calling context
* the element's data is released when the calling context drops it.

I think the actual way to achieve this is to return the `Element` plainly. The calling context allocates it on the stack, then cleans it up.

Reply:
Could be, yes.